File: java-car-data/car-data-web-ui-tests/src/test/java/com/glicerial/samples/cardata/web/uitests/page/EditCarPage.java
package com.glicerial.samples.cardata.web.uitests.page;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;


public class EditCarPage {

    private WebDriver driver;

    @FindBy(id = "year")
    private WebElement yearInput;

    @FindBy(id = "make")
    private WebElement makeInput;

    @FindBy(id = "model")
    private WebElement modelInput;

    @FindBy(id = "trimlevels")
    private WebElement trimLevels;

    @FindBy(id = "addtrimlevelbutton")
    private WebElement addTrimLevelButton;

    @FindBy(id = "submitcarbutton")
    private WebElement submitCarButton;

    public EditCarPage(WebDriver driver) {
        this.driver = driver;
        WebDriverWait wait = new WebDriverWait(driver, 5);
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.id("trimlevels")));
        PageFactory.initElements(driver, this);
    }

    public void clickSubmitCarButton() {
        submitCarButton.click();

        WebDriverWait wait = new WebDriverWait(driver, 5);

        // Dismiss dialog that says "Car updated."
        wait.until(ExpectedConditions.alertIsPresent());
        driver.switchTo().alert().accept();

        // Wait for javascript redirect
        wait.until(ExpectedConditions.presenceOfElementLocated(By.xpath("//div[@class='maincontent']//table[@id='carstable']")));
    }

    public void editYear(String year) {
        yearInput.clear();
        yearInput.sendKeys(year);
    }

    public void editMake(String make) {
        makeInput.clear();
        makeInput.sendKeys(make);
    }

    public void editModel(String model) {
        modelInput.clear();
        modelInput.sendKeys(model);
    }

    public WebElement getTrimLevels() {
        return trimLevels;
    }

    public void clickAddTrimLevelButton() {
        addTrimLevelButton.click();
    }
}

File: python-selenium/pages/edit_car_page.py
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.ui import WebDriverWait

from .locators import EditCarPageLocators, MainPageLocators
from .page import BasePage

class EditCarPage(BasePage):
    def __init__(self, driver):
        super().__init__(driver)

        # Wait for trim levels to be present, others should then also be present
        self.wait = WebDriverWait(self.driver, 10)
        self.trim_levels = self.wait.until(expected_conditions.visibility_of_element_located(EditCarPageLocators.TRIM_LEVELS))

        self.year_input = self.driver.find_element(*EditCarPageLocators.YEAR_INPUT)
        self.make_input = self.driver.find_element(*EditCarPageLocators.MAKE_INPUT)
        self.model_input = self.driver.find_element(*EditCarPageLocators.MODEL_INPUT)

    def edit_year(self, year):
        self.year_input.clear()
        self.year_input.send_keys(year)

    def edit_make(self, make):
        self.make_input.clear()
        self.make_input.send_keys(make)

    def edit_model(self, model):
        self.model_input.clear()
        self.model_input.send_keys(model)

    def click_add_trim_level_button(self):
        add_trim_level_button = self.driver.find_element(*EditCarPageLocators.ADD_TRIM_LEVEL_BUTTON)
        add_trim_level_button.click()

    def click_submit_car_button(self):
        submit_car_button = self.driver.find_element(*EditCarPageLocators.SUBMIT_CAR_BUTTON)
        submit_car_button.click()

        # Accept alert
        self.wait.until(expected_conditions.alert_is_present())
        self.driver.switch_to.alert.accept()

        # Wait for javascript redirect
        self.wait.until(expected_conditions.presence_of_element_located(MainPageLocators.CAR_ROWS))